pp.c

```diff
@@ -7144,6 +7144,25 @@ PP(pp_argcheck)
     return NORMAL;
 }
 
+PP(pp_isa)
+{
+    dSP;
+    SV *left, *right;
+
+    right = POPs;
+    left  = TOPs;
+
+    if(!SvROK(left) || !SvOBJECT(SvRV(left))) {
+        SETs(boolSV(FALSE));
+        RETURN;
+    }
+
+    /* TODO: Support right being an HV ref to the stash directly */
+
+    SETs(boolSV(sv_derived_from_sv(left, right, 0)));
```

Comment:
Is this correct?  Shouldn't it call isa() on the class/object so isa() overrides work?

Reply:
Ahyes, in the general case almost certainly - and definitely for the `DOES` version that will be necessary. I was hoping to avoid that for this initial implementation as it's quite a rare case.

I wonder if it would be possible to optimise it a bit. Attempt to resolve the `"isa"` method and i the likely event that it doesn't exist continue with this optimised path, but if it does exist then invoke that instead via `call_sv()`. I was hoping to avoid a nested code invocation in the common case that nobody overrides `sub isa` though.

Reply:
I've had an attempt at this, but I can't seem to get the nested `call_sv()` call quite right. It mostly runs but then crashes:

http://paste.scsys.co.uk/586753

if I put various warn() calls in there I can see it mostly gets to the right places, or even the presence of warn() makes it work entirely but then if I remove the warn it stops again :( Can anyone suggest what I'm doing wrong here?

Reply:
The answer, for posterity, is that the `SvTRUE()` macro might evaluate its argument more than once, so doing `SvTRUE(POPs)` is always a bug. It should be written `SvTRUEx(POPs)` or maybe `SvTRUE(TOPs); POPs;`. Doing that it works fine.

Reply:
@tonycoz Now added in f0e6ff86a8

Reply:
@xenu  I think it is surely easily possible to find examples on CPAN of people overriding `->isa` to do all sorts of odd things. Whether any of those are "legitimate" is open to question. I'm simply trying to provide a more efficient (notation/implementation) of the same behaviour as is usually achieved by just calling that now.

Besides, there's precedent in the `overload` module to allow all sorts of unary, and binary infix operators between types to be overridden. To not allow this `isa` in similar fashion would seem strange.

Reply:
> Are there any valid reasons to override the `isa` method?

Apparently [Test::MockObject](https://metacpan.org/pod/Test::MockObject) uses an overridden isa.

Reply:
I override isa in [Object::ForkAware](https://metacpan.org/pod/Object::ForkAware) also.

Reply:
Both Moo and Type::Tiny override isa in some places to allow interoperability with Moose.